### PR TITLE
Apple retry uploads

### LINKED ISSUE #685: Retry Apple uploads
The commit https://github.com/PRX/feeder.prx.org/commit/93efce734a4572377f19a1486826f825fed720db was intended handles the case where there is podcast audio uploaded (marked in the container and episode) but the deliveries and delivery files were missing. This was to workaround the fact that if we were using the podcast deliveries and delivery files as sentinels to mark successful upload.

Later, https://github.com/PRX/feeder.prx.org/commit/4678171ebfe775426a910fb031d9f2b6f7d5df1d landed, which effectively marks the episode as "done" so that once audio is successfully uploaded to apple, the [episode never enters into the publishing flow](https://github.com/PRX/feeder.prx.org/blob/230e8527ddabd2a9547aef5567bf94665ae149cc/app/models/apple/publisher.rb#L41).

With https://github.com/PRX/feeder.prx.org/issues/680 it's now clear that using the podcast deliveries as a sentinel at the outset of the publishing flow does not make sense. Instead, we can rely on the [`#synced_with_apple?`](https://github.com/PRX/feeder.prx.org/blob/712f93f0f849d0fc9bde4f4b8e4c60ecdb940f6f/app/models/apple/publisher.rb#L41)  guard to leverage our local state.

If an episode is in the publishing flow, it will be re-processed. So no need to [guard if the delivery is present](https://github.com/PRX/feeder.prx.org/blob/712f93f0f849d0fc9bde4f4b8e4c60ecdb940f6f/app/models/apple/podcast_container.rb#L25).

TODO:
- [x] Set a guard on the Apple::Episode inline with `#synced_with_apple?` but capturing error semantics on the delivery. 
- [x] Filter out episodes with uploaded and processed delivery errors in Apple::Publisher#filter_episodes
- [x] Apple::PodcastContainer resets file source metadata and local delivery state at the outset of the file upload flow, re-uploads audio.

### PR DESCRIPTION
**Pull request by svevang**
closes #685 

This sets up a series of changes to help with upload retries:

- shifts some of the attention from Delivery + DeliveryFile over to the PodcastContainer. The PodcastContainer's `files` attributes is a proxy to the validation and processing status on the Delivery/DeliveryFile models. It appears that [Apple retains the overall container state and discards the delivery state](https://github.com/PRX/feeder.prx.org/issues/680), so likely building transitions in a state machine around the PodcastContainer will be better in the long run.
- Fixes a case where the `duplicate` status in the delivery was treated as an error and not properly covered by the success/error conditionals
- expands a `skip_delivery?` method and related helpers on the container to clarify retry logic.

### CODE REVIEW

File: app/models/apple/episode.rb [hunk 1]
```diff
@@ -439,11 +441,7 @@ def synced_with_apple?
     end
 
     def waiting_for_asset_state?
-      (podcast_delivery_files.length > 0 &&
-        podcast_delivery_files.all?(&:delivered?) &&
-        podcast_delivery_files.all?(&:processed?) &&
-        !podcast_delivery_files.all?(&:processed_errors?) &&
-        !audio_asset_state_finished?)
+      podcast_container.delivery_settled? && !audio_asset_state_finished?
```
Comment (svevang) [inline app/models/apple/episode.rb:444]: so now only looking the episode's `audio_asset_state` and at the podcast container (which manages / coordinates the deliveries).

File: app/models/apple/podcast_container.rb [hunk 2]
```diff
@@ -133,18 +131,14 @@ def self.upsert_podcast_container(episode, row)
     end
 
     def self.get_podcast_containers_via_episodes(api, episodes)
-      # Only query for episodes that don't have a podcast container
-      # The container. Assume that if we have a container record, we don't need to poll.
-      eps_without_container = episodes.filter { |ep| ep.apple_persisted? && ep.missing_container? }
-
```
Comment (svevang) [inline app/models/apple/podcast_container.rb:139]: Trying to move special guards and state checks into more top level concerns (like `PodcastContainer#delivery_settled?`)

File: app/models/apple/publisher.rb [hunk 3]
```diff
@@ -114,6 +114,10 @@ def wait_for_upload_processing(eps)
         pdfs = eps.map(&:podcast_delivery_files).flatten
 
         Apple::PodcastDeliveryFile.wait_for_delivery_files(api, pdfs)
+
+        # Get the latest state of the podcast containers
+        # which should include synced files
+        Apple::PodcastContainer.poll_podcast_container_state(api, eps)
```
Comment (svevang) [inline app/models/apple/publisher.rb:120]: Again, missing local state, this time in the `files` Apple attribute -- which gives a readout of the successfully processed/delivered files.

File: app/models/apple/podcast_delivery_file.rb [hunk 4]
```diff
@@ -327,14 +327,14 @@ def delivered?
     def processed_errors?
       return false unless asset_processing_state.present?
 
-      processed_validation_failed? || processed_duplicate?
```
Comment (svevang) [inline app/models/apple/podcast_delivery_file.rb:330]: Uploading a duplicate file is not really an error.

File: app/models/apple/episode.rb [hunk 5]
```diff
@@ -116,7 +116,7 @@ def self.update_audio_container_reference(api, episodes)
       episode_bridge_results
     end
 
-    def self.remove_audio_container_reference(api, show, episodes)
+    def self.remove_audio_container_reference(api, show, episodes, apple_mark_for_reupload: true)
```
[the remark this answers is not in the document]
Comment (cavis) [reply app/models/apple/episode.rb:119]: +1 i can see that being useful from the ol rails console.

File: app/models/apple/episode.rb [hunk 6]
```diff
@@ -382,9 +386,7 @@ def drafting?
     end
 
     def apple_upload_complete?
-      pdfs = feeder_episode.apple_podcast_delivery_files
-
-      pdfs.present? && pdfs.to_a.flatten.all?(&:apple_complete?)
+      feeder_episode.apple_podcast_container.skip_delivery?
```
Comment (cavis) [inline app/models/apple/episode.rb:389]: Bit odd conceptually (the upload is complete when we're skipping the delivery?)  Though I guess it's all the same - we skip the delivery when the upload is complete.  Same check.

File: app/models/apple/podcast_delivery_file.rb (hunk 4 above)
Comment (cavis) [reply app/models/apple/podcast_delivery_file.rb:330]: +1

**Review by cavis — APPROVED** [review]
👍 looks good to me!

**Pull request merged**